### PR TITLE
check also directories in check_case_conflict

### LINKED ISSUE #70: check-case-conflict doesn't notice directory name conflicts
```
$ git status
On branch master

Initial commit

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

    new file:   TEST/foo
    new file:   test/bar

$ git commit -m "Test"
Check for case conflicts...................................................................................Passed
[master (root-commit) 6931ba5] Test
 2 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 TEST/foo
 create mode 100644 test/bar

$ cat .pre-commit-config.yaml
- repo: git://github.com/pre-commit/pre-commit-hooks
  sha: '0d88124ef6343fbbc6c9d2872853f73546c98a3f'
  hooks:
      - id: check-case-conflict
```

A conflict such as `Test/foo` and `test/foo` is caught though.

### PR DESCRIPTION
**Pull request by slsyy**
fix #70

### CODE REVIEW

File: pre_commit_hooks/check_case_conflict.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import argparse
+import pathlib
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:2]: I would prefer if you don't use pathlib, thanks!

File: pre_commit_hooks/check_case_conflict.py [hunk 2]
```diff
@@ -12,9 +13,20 @@ def lower_set(iterable: Iterable[str]) -> Set[str]:
...
+    return files | directories
+
+
 def find_conflicting_filenames(filenames: Sequence[str]) -> int:
-    repo_files = set(cmd_output('git', 'ls-files').splitlines())
-    relevant_files = set(filenames) | added_files()
+    repo_files = add_directories(
+        set(cmd_output('git', 'ls-files').splitlines()),
+    )
+    relevant_files = add_directories(set(filenames) | added_files())
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:29]: probably easier to read as:

```pyhton
repo_files = ...
repo_files |= directories_for(repo_files)
relevant_files = ...
relevant_files |= directories_for(relevant_files)
```

this'll also reduce the amount of set churn

File: pre_commit_hooks/check_case_conflict.py [hunk 3]
```diff
@@ -31,8 +43,7 @@ def find_conflicting_filenames(filenames: Sequence[str]) -> int:
 
     if conflicts:
         conflicting_files = [
-            x for x in repo_files | relevant_files
-            if x.lower() in conflicts
+            x for x in repo_files | relevant_files if x.lower() in conflicts
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:46]: no idea why this line changed, please leave the original

File: pre_commit_hooks/check_case_conflict.py [hunk 4]
```diff
@@ -44,7 +55,8 @@ def find_conflicting_filenames(filenames: Sequence[str]) -> int:
 def main(argv: Optional[Sequence[str]] = None) -> int:
     parser = argparse.ArgumentParser()
     parser.add_argument(
-        'filenames', nargs='*',
+        'filenames',
+        nargs='*',
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:59]: this seems irrelevant to your change

File: tests/check_case_conflict_test.py [hunk 5]
```diff
@@ -46,6 +73,18 @@ def test_file_conflicts_with_committed_file(temp_git_dir):
         assert find_conflicting_filenames(['F.py']) == 1
 
 
+def test_file_conflicts_with_committed_dir(temp_git_dir):
+    with temp_git_dir.as_cwd():
+        temp_git_dir.mkdir('dir').join('x').write('foo')
+        cmd_output('git', 'add', '-A')
+        cmd_output('git', 'commit', '--no-gpg-sign', '-n', '-m', 'Add f.py')
+
+        temp_git_dir.join('DIR').write('foo')
```
Comment (asottile) [inline tests/check_case_conflict_test.py:82]: this is failing on windows -- I don't think you need to create the actual filesystem here (?)

File: pre_commit_hooks/check_case_conflict.py (hunk 1 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:2]: is there any specific reason?

File: pre_commit_hooks/check_case_conflict.py (hunk 2 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:29]: ok

File: pre_commit_hooks/check_case_conflict.py (hunk 3 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:46]: it is formatted by the `pre-commit`

File: pre_commit_hooks/check_case_conflict.py (hunk 4 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:59]: it is formatted by the `pre-commit`

File: pre_commit_hooks/check_case_conflict.py (hunk 3 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:46]: will revert

File: pre_commit_hooks/check_case_conflict.py (hunk 4 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:59]: will revert

File: tests/check_case_conflict_test.py (hunk 5 above)
Comment (slsyy) [reply tests/check_case_conflict_test.py:82]: the `check_case_conflict` use `git ls-files` to list all files in a repo, so I don't see any way that emulate this without touching a filesystem. My first think was to disable these tests on windows, as with that check we want to prevent this from happening on systems, which are case sensitive

File: pre_commit_hooks/check_case_conflict.py (hunk 1 above)
Comment (asottile) [reply pre_commit_hooks/check_case_conflict.py:2]: you can accomplish the same without it, and I don't like pathlib

File: tests/check_case_conflict_test.py (hunk 5 above)
Comment (asottile) [reply tests/check_case_conflict_test.py:82]: hmmm yeah maybe that's fine -- could also emulate the output using some mocking tricks, up to you though

**Comment by slsyy** [comment]
@asottile please re-review

File: pre_commit_hooks/check_case_conflict.py [hunk 6]
```diff
@@ -12,9 +13,23 @@ def lower_set(iterable: Iterable[str]) -> Set[str]:
     return {x.lower() for x in iterable}
 
 
+def parents(file: str) -> Iterator[str]:
+    while True:
+        file = file[:file.rfind('/')]
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:18]: I believe you can use `os.path.dirname` for this

File: pre_commit_hooks/check_case_conflict.py [hunk 7]
```diff
@@ -12,9 +13,23 @@ def lower_set(iterable: Iterable[str]) -> Set[str]:
...
+        if not file:
+            return
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:20]: this can be `while file != os.sep:` I think ?

**Review by asottile — COMMENTED** [review]
looks good, just a few small things -- sorry for the delay I've been very busy

File: pre_commit_hooks/check_case_conflict.py (hunk 7 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:20]: - `os.sep` is different on windows, git always return paths in a posix format
- paths looks like that one: `testing/resources/does_exist`. You don't have a leading `/`

Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py:21]: done, it seems to work

File: pre_commit_hooks/check_case_conflict.py (hunk 6 above)
Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py]: done, it seems to work

File: pre_commit_hooks/check_case_conflict.py [hunk 8]
```diff
@@ -12,9 +14,23 @@ def lower_set(iterable: Iterable[str]) -> Set[str]:
     return {x.lower() for x in iterable}
 
 
+def parents(file: str) -> Iterator[str]:
+    while True:
+        file = os.path.dirname(file)
+        if not file:
+            return
+        yield file
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py:22]:
```python
file = os.path.dirname(file)
while file:
    yield file
    file = os.path.dirname(file)
```

Comment (slsyy) [reply pre_commit_hooks/check_case_conflict.py]: done

**Review by asottile — APPROVED** [review]
